drivers/usb/udc/udc_bflb_bl61x.c

```diff
@@ -0,0 +1,1480 @@
+	sys_write32(tmp, PDS_BASE + PDS_USB_PHY_CTRL_OFFSET);
+
+	tmp = sys_read32(PDS_BASE + PDS_USB_PHY_CTRL_OFFSET);
+	tmp &= ~PDS_REG_USB_PHY_PONRST_MSK;
+	sys_write32(tmp, PDS_BASE + PDS_USB_PHY_CTRL_OFFSET);
+
+	tmp = sys_read32(PDS_BASE + PDS_USB_CTL_OFFSET);
+	tmp &= ~PDS_REG_USB_EXT_SUSP_N_MSK;
+	sys_write32(tmp, PDS_BASE + PDS_USB_CTL_OFFSET);
+
+	return 0;
+}
+
+static int udc_bflb_bl61x_driver_preinit(const struct device *dev)
+{
+	const struct udc_bflb_bl61x_config *config = dev->config;
+	struct udc_data *data = dev->data;
+	struct udc_bflb_bl61x_data *priv = data->priv;
+	uint16_t mps = 64;
+	int err;
+
+	k_mutex_init(&data->mutex);
+	k_fifo_init(&priv->fifo);
+
+	data->caps.rwup = true;
+	data->caps.mps0 = UDC_MPS0_64;
+	if (config->speed_idx == UDC_BUS_SPEED_HS) {
+		data->caps.hs = true;
+		mps = 1024;
+	}
+
+	for (int i = 0; i < config->num_of_eps; i++) {
+		config->ep_cfg_out[i].caps.out = 1;
+		if (i == 0) {
+			config->ep_cfg_out[i].caps.control = 1;
+			config->ep_cfg_out[i].caps.mps = 64;
+		} else {
+			config->ep_cfg_out[i].caps.bulk = 1;
+			config->ep_cfg_out[i].caps.interrupt = 1;
+			config->ep_cfg_out[i].caps.iso = 1;
+			config->ep_cfg_out[i].caps.mps = mps;
+		}
+
+		config->ep_cfg_out[i].addr = USB_EP_DIR_OUT | i;
+		err = udc_register_ep(dev, &config->ep_cfg_out[i]);
+		if (err != 0) {
+			LOG_ERR("Failed to register endpoint");
+			return err;
+		}
+	}
+
+	for (int i = 0; i < config->num_of_eps; i++) {
+		config->ep_cfg_in[i].caps.in = 1;
+		if (i == 0) {
+			config->ep_cfg_in[i].caps.control = 1;
+			config->ep_cfg_in[i].caps.mps = 64;
+		} else {
+			config->ep_cfg_in[i].caps.bulk = 1;
+			config->ep_cfg_in[i].caps.interrupt = 1;
+			config->ep_cfg_in[i].caps.iso = 1;
+			config->ep_cfg_in[i].caps.mps = mps;
+		}
+
+		config->ep_cfg_in[i].addr = USB_EP_DIR_IN | i;
+		err = udc_register_ep(dev, &config->ep_cfg_in[i]);
+		if (err != 0) {
+			LOG_ERR("Failed to register endpoint");
+			return err;
+		}
+	}
+
+	LOG_INF("Device %p (max. speed %d)", dev, config->speed_idx);
+
+	return 0;
+}
+
+static void udc_bflb_bl61x_lock(const struct device *dev)
+{
+	udc_lock_internal(dev, K_FOREVER);
+}
+
+static void udc_bflb_bl61x_unlock(const struct device *dev)
+{
+	udc_unlock_internal(dev);
+}
+
+static void udc_bflb_bl61x_isr(const struct device *dev)
+{
+	const struct udc_bflb_bl61x_config *cfg = dev->config;
+	struct udc_bflb_bl61x_data *priv = udc_get_private(dev);
+	uint32_t glb_intstatus;
+	uint32_t dev_intstatus;
+	uint32_t group_intstatus;
+	uint32_t tmp;
+;
+	glb_intstatus = sys_read32(cfg->base + USB_GLB_ISR_OFFSET);
+
+	if (glb_intstatus & USB_DEV_INT) {
+		dev_intstatus = sys_read32(cfg->base + USB_DEV_IGR_OFFSET);
+		if (dev_intstatus & USB_INT_G0) {
+			group_intstatus = sys_read32(cfg->base + USB_DEV_ISG0_OFFSET);
+			group_intstatus &=  ~sys_read32(cfg->base + USB_DEV_MISG0_OFFSET);
+
+			if (group_intstatus & USB_CX_COMABT_INT) {
+				udc_submit_event(dev, UDC_EVT_ERROR, -1);
```

Comment:
Maybe `-ECANCELED` is fitting?
```suggestion
				udc_submit_event(dev, UDC_EVT_ERROR, -ECANCELED);
```

Reply:
done